src/algebraic_geometry/structure_sheaf.lean

```diff
@@ -0,0 +1,233 @@
+We define this as a subsheaf of the sheaf of dependent functions into the localizations,
+cut out by the condition that the function must be locally equal to a ratio of elements of `R`.
+
+Because the condition "is equal to a fraction" passes to smaller open subsets,
+the subset of functions satisfying this condition is automatically a subpresheaf.
+Because the condition "is locally equal to a fraction" is local,
+it is also a subsheaf.
+
+We also set up the ring structure, obtaining
+`structure_sheaf R : sheaf CommRing (Top.of (prime_spectrum R))`.
+-/
+
+universe u
+
+noncomputable theory
+
+variables (R : Type u) [comm_ring R]
+
+open Top
+open topological_space
+open category_theory
+open opposite
+
+namespace algebraic_geometry
+
+namespace structure_sheaf
+
+/--
+The type family over `prime_spectrum R` consisting of the localization over each point.
+-/
+@[derive [comm_ring, local_ring]]
+def localizations (P : Top.of (prime_spectrum R)) : Type u := localization.at_prime P.as_ideal
+
+instance (P : Top.of (prime_spectrum R)) : inhabited (localizations R P) :=
+⟨(localization.of _).to_map 1⟩
+
+variables {R}
+
+/--
+The predicate saying that a dependent function on an open `U` is realised as a fixed fraction
+`r / s` in each of the stalks (which are localizations at various prime ideals).
+-/
+def is_fraction {U : opens (Top.of (prime_spectrum R))} (f : Π x : U, localizations R x) : Prop :=
+∃ (r s : R), ∀ x : U,
+  ¬ (s ∈ x.1.as_ideal) ∧ f x * (localization.of _).to_map s = (localization.of _).to_map r
+
+variables (R)
+
+/--
+The predicate `is_fraction` is "prelocal",
+in the sense that if it holds on `U` it holds on any open subset `V` of `U`.
+-/
+def is_fraction_prelocal : prelocal_predicate (localizations R) :=
+{ pred := λ U f, is_fraction f,
+  res := by { rintro V U i f ⟨r, s, w⟩, exact ⟨r, s, λ x, w (i x)⟩ } }
+
+variables {R}
+
+/--
+We will define the structure sheaf as
+the subsheaf of all dependent functions in `Π x : U, localizations R x`
+consisting of those functions which can locally be expressed as a ratio of
+(the images in the localization of) elements of `R`.
+
+Quoting Hartshorne:
+
+For an open set $$U ⊆ Spec A$$, we define $$𝒪(U)$$ to be the set of functions
+$$s : U → ⨆_{𝔭 ∈ U} A_𝔭$$, such that $s(𝔭) ∈ A_𝔭$$ for each $$𝔭$$,
+and such that $$s$$ is locally a quotient of elements of $$A$$:
+to be precise, we require that for each $$𝔭 ∈ U$$, there is a neighborhood $$V$$ of $$𝔭$$,
+contained in $$U$$, and elements $$a, f ∈ A$$, such that for each $$𝔮 ∈ V, f ∉ 𝔮$$,
+and $$s(𝔮) = a/f$$ in $$A_𝔮$$.
+
+Now Hartshorne had the disadvantage of not knowing about dependent functions,
+so we replace his circumlocution about functions into a disjoint union with
+`Π x : U, localizations x`.
+-/
+def is_locally_fraction
+  {U : opens (Top.of (prime_spectrum R))} (f : Π x : U, localizations R x) : Prop :=
+∀ x : U, ∃ (V) (m : x.1 ∈ V) (i : V ⟶ U),
+  ∃ (r s : R), ∀ y : V,
```

Comment:
Isn't this just `is_fraction (res V f)`? Is there a reason you want to write the definition out again?

Reply:
Yes, it is. It's written out again here as a historical accident --- we wrote this before we even implemented `prelocal_predicate.sheafify`. I guess I should just nuke it and streamline things. I'll get to it in a few hours.

Reply:
Okay, I've restructured this so it's more obvious that when we write it out again it's as a `rfl` lemma, not an independent definition.